resharper/resharper-unity/src/CSharp/Daemon/Stages/Analysis/MultidimensionalArraysDeclarationProblemAnalyzer.cs

```diff
@@ -0,0 +1,89 @@
+using System.Linq;
+using JetBrains.Annotations;
+using JetBrains.ReSharper.Daemon.CSharp.Stages;
+using JetBrains.ReSharper.Feature.Services.Daemon;
+using JetBrains.ReSharper.Plugins.Unity.CSharp.Daemon.Errors;
+using JetBrains.ReSharper.Plugins.Unity.CSharp.Daemon.Stages.Dispatcher;
+using JetBrains.ReSharper.Psi;
+using JetBrains.ReSharper.Psi.CSharp;
+using JetBrains.ReSharper.Psi.CSharp.Tree;
+using JetBrains.ReSharper.Psi.Pointers;
+using JetBrains.ReSharper.Psi.Tree;
+
+namespace JetBrains.ReSharper.Plugins.Unity.CSharp.Daemon.Stages.Analysis
+{
+    [ElementProblemAnalyzer(typeof(IVariableDeclaration), typeof(InefficientMultidimensionalArrayUsageWarning))]
+    public class MultidimensionalArraysDeclarationProblemAnalyzer : UnityElementProblemAnalyzer<IVariableDeclaration>
+    {
+        public MultidimensionalArraysDeclarationProblemAnalyzer([NotNull] UnityApi unityApi)
+            : base(unityApi)
+        {
+        }
+
+        protected override void Analyze(IVariableDeclaration element, ElementProblemAnalyzerData data, IHighlightingConsumer consumer)
+        {
+            var declaredElement = element.DeclaredElement;
+            if (declaredElement == null)
+                return;
+
+            var classDeclaration = element.GetContainingNode<IClassLikeDeclaration>();
+            if (classDeclaration == null)
+                return;
+
+            var type = element.Type as IArrayType;
+            if (type == null)
+                return;
+            
+            if (type.Rank == 1)
+                return;
+            
+            IArrayCreationExpression arrayCreationExpression = null;
+            IMultipleDeclaration multipleDeclaration = null;
+
+            switch (element)
+            {
+                case IFieldDeclaration fieldDeclaration:
+                    if (fieldDeclaration.GetAccessRights() != AccessRights.PRIVATE)
```

Comment:
Why only `private` fields?

Reply:
Performance reasons, we could guaranty that we will not break semantic after quickfix only for local members